### DISCUSSION TITLE
replay speed / position

### DISCUSSION THREAD

**Discussion by ThomasCr** [stronnag/mwptools#171] [Ideas]
Hi,

it would be nice to get a slider for blackbox replays to go back or jump forward some seconds. Also it would be nice to set a replay speed for analysis.

Many Thanks

**Comment by stronnag** [comment]
Nice idea, just needs someone to implement it (non-trivial in the current architecture).

**Comment by stronnag** [comment]
This has actually been bugging me for at least 9 years; so much so I finally decided to do something about it.
Still work in progress (no branch yet) and it will require a new version of `bbl2kml`/ `flightlog2kml`. 
* BBL files (and OTX/ETX) logs are inherently sequential read (vice random access).
* Startup time will be a bit slower as the whole log is read to a SQLite database
* The SQLite DB is then used to for the replay. We now have an easily seekable (get data from any point) source.
* Some feature of the extant replay will not be implemented at first (video linking, mission "OSD").

Teaser:

https://github.com/user-attachments/assets/f16c8aec-6cf0-4da1-81cf-54b122e3d001

**Comment by stronnag** [comment]
## Test branch

There is now a repository branch that includes the new log replayer.

* You must first install the **development** branch of [flightlog2kml](https://github.com/stronnag/bbl2kml)
```
  $ flightlog2kml --version
  flightlog2kml 1.0.30 commit:08a0e34
```
* Checkout and build / install the **sqlbbl** branch of mwptools

There is  a new menu option **"Experimental log player"**. This loads a somewhat familiar chooser for Blackbox and OTX/ETX logs.
* Selecting a log entry and clicking "OK" will present the player; if the (default) "Autostart" chooser option was enabled, then replay starts immediately, otherwise the replay can be manually started.
* The replay may be paused / advanced / retarded / restarted using the play button / slider / goto start / goto end buttons.
* Closing the slider window terminates the replay.

### Other notes

* Some feature of the extant replay will not be implemented at first (video linking, mission "OSD").
* The "Flight Statistics" dialog is updated whenever the replay is paused (or closed). It always details for  the whole flight rather than incremental  values.
* Direction (GGS v. mag) analysis is not yet implemented.
* The new chooser / player will be used for command line` -b, --replay-bbox` and drag / drop files on to the mwp window if the environment variable `MWP_PREFER_XLOG` is set (to anything), e.g:
```
  $ MWP_PREFER_XLOG=1 mwp -b /t/inav-contrib/LOG00006.TXT
```

### Reporting Issues

Please include:

* The flightlog
* The mwp "stderr" log
* A comprehensive description of how to reproduce / expected behaviour etc.

**Comment by stronnag** [comment]
Menu option renamed ...
<img width="299" height="196" alt="image" src="https://github.com/user-attachments/assets/203248ab-ca3a-471c-9771-d25b6007af09" />

And legacy items in sub-menu.
<img width="449" height="200" alt="image" src="https://github.com/user-attachments/assets/79f8329f-d327-4c6b-9940-3ce6115632d6" />

If the log file contains multiple logs,  there will be a player menu so one can switch logs without having to re-parse the whole file ...

<img width="490" height="269" alt="image" src="https://github.com/user-attachments/assets/f30d9c45-4730-4e65-b6f0-7924bc1e3b98" />

**Reply by MiguelitoBeard** [reply]
This is really interesting! Thanks for all the work!

**Comment by stronnag** [comment]
now mwp development branch (and bbl2kml master / 1.0.30)

**Comment by MiguelitoBeard** [comment]
@ThomasCr let me know how it goes and I will follow :)

**Comment by stronnag** [comment]
This was included in [25.07.22](https://github.com/stronnag/mwptools/releases/tag/25.07.22).  [User Guide](https://www.daria.co.uk/mwptools/replay-tools/#interactive-player)

Known Issues:

* Slider movement on fast-forward is disconnected from track generation. This is fixed in `development` branch.
* The generated SQLite files are left in `$TMPDIR` (`/tmp` on POSIX). This may (or may not) be considered a bug.
